File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/witchwater/WitchWaterWorldCategory.java
package wraith.fabricaeexnihilo.compatibility.rei.witchwater;

import me.shedaniel.math.Point;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.Renderer;
import me.shedaniel.rei.api.client.gui.widgets.Widget;
import me.shedaniel.rei.api.client.gui.widgets.Widgets;
import me.shedaniel.rei.api.client.registry.display.DisplayCategory;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.text.Text;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.compatibility.rei.GlyphWidget;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;
import wraith.fabricaeexnihilo.modules.witchwater.WitchWaterFluid;

import java.util.ArrayList;
import java.util.List;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public class WitchWaterWorldCategory implements DisplayCategory<WitchWaterWorldDisplay> {

    public static final int GLYPH_ARROW_U = 0;
    public static final int GLYPH_ARROW_V = 0;
    public static final int GLYPH_PLUS_U = 3 * 16;
    public static final int GLYPH_PLUS_V = 0;
    public static final int MARGIN = 6;
    public static final int GLYPH_PLUS_X = MARGIN;
    public static final int GLYPH_ARROW_X = MARGIN + 18;
    public static final int OUTPUT_X = GLYPH_ARROW_X + 18;
    public static final int FLUID_X = MARGIN;
    public static final int OUTPUT_SLOTS_X = Math.max(FabricaeExNihilo.CONFIG.modules.REI.witchwaterworldCols, 1);
    public static final int OUTPUT_SLOTS_Y = Math.max(FabricaeExNihilo.CONFIG.modules.REI.witchwaterworldRows, 3);
    public static final int MAX_OUTPUTS = OUTPUT_SLOTS_X * OUTPUT_SLOTS_Y;
    public static final int HEIGHT = OUTPUT_SLOTS_Y * 18 + MARGIN * 2;
    public static final int GLYPH_PLUS_Y = HEIGHT / 2 - 9;
    public static final int GLYPH_ARROW_Y = GLYPH_PLUS_Y;
    public static final int FLUID_Y = GLYPH_PLUS_Y + 18;
    public static final int WITCH_Y = GLYPH_PLUS_Y - 18;
    public static final int OUTPUT_Y = MARGIN;
    public static final int WIDTH = (OUTPUT_SLOTS_X + 2) * 18 + MARGIN * 2;
    public static final int WITCH_X = MARGIN;
    public static Identifier GLYPHS = id("textures/gui/rei/glyphs.png");

    @Override
    public CategoryIdentifier<? extends WitchWaterWorldDisplay> getCategoryIdentifier() {
        return PluginEntry.WITCH_WATER_WORLD;
    }

    @Override
    public int getDisplayHeight() {
        return HEIGHT;
    }

    @Override
    public int getDisplayWidth(WitchWaterWorldDisplay display) {
        return WIDTH;
    }

    @Override
    public Renderer getIcon() {
        return EntryStacks.of(WitchWaterFluid.BUCKET);
    }

    @Override
    public Text getTitle() {
        return new TranslatableText("fabricaeexnihilo.rei.category.witch_water.world");
    }

    @Override
    public List<Widget> setupDisplay(WitchWaterWorldDisplay display, Rectangle bounds) {
        var widgets = new ArrayList<Widget>();
        widgets.add(Widgets.createRecipeBase(bounds));

        var glyphPlus = new GlyphWidget(bounds, bounds.getMinX() + GLYPH_PLUS_X, bounds.getMinY() + GLYPH_PLUS_Y, 16, 16, GLYPHS, GLYPH_PLUS_U, GLYPH_PLUS_V);
        var glyphArrow = new GlyphWidget(bounds, bounds.getMinX() + GLYPH_ARROW_X, bounds.getMinY() + GLYPH_ARROW_Y, 16, 16, GLYPHS, GLYPH_ARROW_U, GLYPH_ARROW_V);
        widgets.add(glyphPlus);
        widgets.add(glyphArrow);

        var witchWater = EntryIngredients.of(WitchWaterFluid.STILL);
        var fluids = display.getInputEntries().get(0);
        var results = display.getOutputEntries();

        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + WITCH_X, bounds.getMinY() + WITCH_Y)).entries(witchWater));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + FLUID_X, bounds.getMinY() + FLUID_Y)).entries(fluids));

        for (int y = 0; y < MAX_OUTPUTS; ++y) {
            var slot = Widgets.createSlot(new Point(bounds.getMinX() + OUTPUT_X + (y % OUTPUT_SLOTS_X) * 18, bounds.getMinY() + OUTPUT_Y + (y / OUTPUT_SLOTS_X) * 18));
            if (y < results.size()) {
                slot.entries(results.get(y));
            }
            widgets.add(slot);
        }

        return widgets;
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/witchwater/WitchWaterWorldCategory.kt
package exnihilofabrico.compatibility.rei.witchwater

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.compatibility.rei.GlyphWidget
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.id
import exnihilofabrico.modules.witchwater.WitchWaterFluid
import exnihilofabrico.util.asREIEntry
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.api.RecipeCategory
import me.shedaniel.rei.gui.widget.EntryWidget
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import me.shedaniel.rei.gui.widget.Widget
import java.util.function.Supplier

class WitchWaterWorldCategory: RecipeCategory<WitchWaterWorldDisplay> {

    override fun getIdentifier() = PluginEntry.WITCH_WATER_WORLD
    override fun getLogo() = WitchWaterFluid.bucket.asREIEntry()
    override fun getCategoryName() = "Witch Water Fluid Interactions"


    override fun getDisplayHeight() = HEIGHT
    override fun getDisplayWidth(display: WitchWaterWorldDisplay) = WIDTH

    override fun setupDisplay(displaySupplier: Supplier<WitchWaterWorldDisplay>, bounds: Rectangle): MutableList<Widget> {
        val display = displaySupplier.get()
        val widgets = mutableListOf<Widget>(RecipeBaseWidget(bounds))

        val glyphPlus = GlyphWidget(bounds, bounds.minX + GLYPH_PLUS_X, bounds.minY + GLYPH_PLUS_Y, GLYPH_WIDTH, GLYPH_HEIGHT, GLYPHS, GLYPH_PLUS_U, GLYPH_PLUS_V)
        val glypgArrow = GlyphWidget(bounds, bounds.minX + GLYPH_ARROW_X, bounds.minY + GLYPH_ARROW_Y, GLYPH_WIDTH, GLYPH_HEIGHT, GLYPHS, GLYPH_ARROW_U, GLYPH_ARROW_V)
        widgets.add(glyphPlus)
        widgets.add(glypgArrow)

        val witchWaterwitchWater = display.inputEntries[0]
        val fluids = display.inputEntries[1]
        val results = display.outputEntries

        widgets.add(EntryWidget.create(bounds.minX + WITCH_X, bounds.minY + WITCH_Y).entries(witchWaterwitchWater))
        widgets.add(EntryWidget.create(bounds.minX + FLUID_X, bounds.minY + FLUID_Y).entries(fluids))


        results.forEachIndexed { index, result ->
            widgets.add(
                EntryWidget.create(
                    bounds.minX + OUTPUT_X + (index % OUTPUT_SLOTS_X)*18,
                    bounds.minY + OUTPUT_Y + (index / OUTPUT_SLOTS_X)*18).entry(result))
        }
        // Fill in the empty spots
        for(index in results.size until MAX_OUTPUTS) {
            widgets.add(
                EntryWidget.create(
                    bounds.minX + OUTPUT_X + (index % OUTPUT_SLOTS_X)*18,
                    bounds.minY + OUTPUT_Y + (index / OUTPUT_SLOTS_X)*18))
        }

        return widgets
    }

    companion object {
        val GLYPHS = id("textures/gui/rei/glyphs.png")

        val OUTPUT_SLOTS_X = maxOf(ExNihiloFabrico.config.modules.REI.witchwaterworldCols, 1)
        val OUTPUT_SLOTS_Y = maxOf(ExNihiloFabrico.config.modules.REI.witchwaterworldRows, 3)

        val MAX_OUTPUTS = OUTPUT_SLOTS_X*OUTPUT_SLOTS_Y

        val MARGIN = 6

        val WIDTH = (OUTPUT_SLOTS_X+2)*18 + MARGIN*2
        val HEIGHT = OUTPUT_SLOTS_Y*18 + MARGIN*2

        val GLYPH_PLUS_X = MARGIN
        val GLYPH_PLUS_Y = HEIGHT/2 - 9
        val GLYPH_ARROW_X = MARGIN + 18
        val GLYPH_ARROW_Y= GLYPH_PLUS_Y

        val OUTPUT_X = GLYPH_ARROW_X + 18
        val OUTPUT_Y = MARGIN

        val FLUID_X = MARGIN
        val FLUID_Y = GLYPH_PLUS_Y + 18
        val WITCH_X = MARGIN
        val WITCH_Y = GLYPH_PLUS_Y - 18

        val GLYPH_WIDTH = 16
        val GLYPH_HEIGHT = 16
        val GLYPH_PLUS_U = 3*16
        val GLYPH_PLUS_V = 0
        val GLYPH_ARROW_U = 0
        val GLYPH_ARROW_V= 0
    }

}